File: go/152.go

//package main

//import "fmt"

// 6 star, no idea, 由于负负得正，所以负数的结果也要记录
// positive = max(num, positive * num, negative * num)
// negative = min(num, temp * num, negative * num)
// go里没有直接提供max min这样的函数，所以代码里写的比较冗长
func maxProduct(nums []int) int {
	var temp int
	positive, negative,  max := nums[0], nums[0], nums[0]
	for _, val:= range nums[1:] {
		positive *= val
		negative *= val
		if positive < negative {
			temp = positive
			positive = negative
			negative = temp
		}
		if positive < val {
			positive = val
		}
		if negative > val {
			negative = val
		}

		if positive > max {max = positive}
	}
	return max
}


//func main() {
//	nums := []int {2,3,-2,4}
//	ret := maxProduct(nums)
//	fmt.Println(ret)
//}

File: v2/152.py



class Solution:
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # 6 star, 没做出来
        # 从第二个开始，每次记录最大值和最小值，因为最小值乘以一个负数也可能是最大值
        _max = small = big =  nums[0]
        for i in nums[1:]:
            # small和big的递推公式
            small, big = min(i, i*big, i*small), max(i, i*big, i*small)
            _max = max(_max, big)
        return _max

